src/fsharp/FSharp.Core/reflect.fs

```diff
@@ -374,148 +374,63 @@ module internal Impl =
 
     let maxTuple = 8
     // Which field holds the nested tuple?
     let tupleEncField = maxTuple-1
 
-    let rec mkTupleType isStruct (tys: Type[]) =
-        let deOption (o: option<Type>) =
-            match o with 
-            | Some v -> v
-            | None -> raise systemValueTupleException
-        if isStruct then
-            match tys.Length with 
-            | 1 -> deOption(stuple1).MakeGenericType(tys)
-            | 2 -> deOption(stuple2).MakeGenericType(tys)
-            | 3 -> deOption(stuple3).MakeGenericType(tys)
-            | 4 -> deOption(stuple4).MakeGenericType(tys)
-            | 5 -> deOption(stuple5).MakeGenericType(tys)
-            | 6 -> deOption(stuple6).MakeGenericType(tys)
-            | 7 -> deOption(stuple7).MakeGenericType(tys)
-            | n when n >= maxTuple -> 
-                let tysA = tys.[0..tupleEncField-1]
-                let tysB = tys.[maxTuple-1..]
-                let tyB = mkTupleType isStruct tysB
-                deOption(stuple8).MakeGenericType(Array.append tysA [| tyB |])
-            | _ -> invalidArg "tys" (SR.GetString(SR.invalidTupleTypes))
-        else
-            match tys.Length with 
-            | 1 -> tuple1.MakeGenericType(tys)
-            | 2 -> tuple2.MakeGenericType(tys)
-            | 3 -> tuple3.MakeGenericType(tys)
-            | 4 -> tuple4.MakeGenericType(tys)
-            | 5 -> tuple5.MakeGenericType(tys)
-            | 6 -> tuple6.MakeGenericType(tys)
-            | 7 -> tuple7.MakeGenericType(tys)
-            | n when n >= maxTuple ->
-                let tysA = tys.[0..tupleEncField-1]
-                let tysB = tys.[maxTuple-1..]
-                let tyB = mkTupleType isStruct tysB
-                tuple8.MakeGenericType(Array.append tysA [| tyB |])
-            | _ -> invalidArg "tys" (SR.GetString(SR.invalidTupleTypes))
-
-    let rec getTupleTypeInfo    (typ:Type) = 
+    let rec mkTupleType isStruct (asm:Assembly) (tys: Type[]) =
+        let tupleFullName n =
+            if isStruct then
+                match n with
+                | 1 -> "System.ValueTuple`1"
+                | 2 -> "System.ValueTuple`2"
+                | 3 -> "System.ValueTuple`3"
+                | 4 -> "System.ValueTuple`4"
+                | 5 -> "System.ValueTuple`5"
+                | 6 -> "System.ValueTuple`6"
+                | 7 -> "System.ValueTuple`7"
+                | 8 -> "System.ValueTuple`8"
+                | _ -> "System.ValueTuple"
+            else
+                match n with
+                | 1 -> "System.Tuple`1"
+                | 2 -> "System.Tuple`2"
+                | 3 -> "System.Tuple`3"
+                | 4 -> "System.Tuple`4"
+                | 5 -> "System.Tuple`5"
+                | 6 -> "System.Tuple`6"
+                | 7 -> "System.Tuple`7"
+                | 8 -> "System.Tuple`8"
+                | _ -> "System.Tuple"
+
+        match tys.Length with 
+        | 1 -> asm.GetType(tupleFullName 1).MakeGenericType(tys)
```

Comment:
I'm a bit concerned the performance of ``MakeTupleType`` has degraded considerably because of all these string constructions and name lookups.  I could imagine it being used fairly intensively by some reflection-based deserializers, for exampe. Finding a way to amortize these would be sensible I think.  Or at least measuring the perf difference.